### PR TITLE
Migrate rich text utils to TS

### PR DESCRIPTION
**Pull request by ddouglasz**
#### Summary

- Migrate rich-text-util to typescript

### CODE REVIEW

File: packages/components/inputs/rich-text-utils/src/index.ts [hunk 1]
```diff
@@ -2,6 +2,7 @@ export * as localized from './localized';
 export { default as html } from './html';
 export { default as isEmpty } from './is-empty';
 export { default as richTextPlugins } from './rich-text-plugins';
+// eslint-disable-next-line import/no-unresolved
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/index.ts:5]: I don't think we are supposed to have this eslint comment

Comment (Rhotimee) [reply packages/components/inputs/rich-text-utils/src/index.ts:5]: I think the reason you're getting the error is because of the file name. It is `rich-text-body/index.Ts` instead of `rich-text-body/index.ts`

**Review by Rhotimee — CHANGES_REQUESTED** [review]

File: packages/components/inputs/rich-text-utils/src/index.ts (hunk 1 above)
Comment (ddouglasz) [reply packages/components/inputs/rich-text-utils/src/index.ts:5]: Ah! thank you. Nice catch.

File: packages/components/inputs/rich-text-utils/src/localized/localized.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      }
+      return localizedStringWithoutEmptyTranslations;
+    },
+    {}
+  );
+};
+
+export const createLocalizedString = (
+  languages: string[],
+  existingTranslations = {} as TCreateLocalizedString
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/localized/localized.ts:50]: As the type is typing a single parameter I was expecting it to have a name resembling the expected value; something like `TTranslations`.
What do you think?

File: packages/components/inputs/rich-text-utils/src/plugins/mark.tsx [hunk 3]
```diff
@@ -1,6 +1,40 @@
 import { isKeyHotkey } from 'is-hotkey';
 import { warning } from '@commercetools-uikit/utils';
 import memoize from 'lodash/memoize';
+import { ReactNode } from 'react';
+import { ReactJSXElement } from '@emotion/react/types/jsx-namespace';
+
+type TOptions = {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/plugins/mark.tsx:7]: Maybe a name like `TMarkPluginOptions` would be more specific and relatable to its use.

Comment (ddouglasz) [reply packages/components/inputs/rich-text-utils/src/plugins/mark.tsx:7]: Thanks Carlos. Will work on all the feedback

File: packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.tsx [hunk 4]
```diff
@@ -10,26 +14,69 @@ import {
...
+  };
+
+  const Downshift = DownshiftUntyped as ElementType;
 
   return (
     <Downshift
       onChange={props.onChange}
       selectedItem={props.value}
-      itemToString={(headings) => (headings ? headings.label : '')}
+      itemToString={(headings: THeadings) => (headings ? headings?.label : '')}
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.tsx:67]: nit: Now that we're using the optional chaining, it seems we don't need the ternary anymore:
```suggestion
      itemToString={(headings: THeadings) => (headings?.label || '')}
```

File: packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.tsx [hunk 5]
```diff
@@ -61,13 +108,13 @@ const Dropdown = (props) => {
...
                 <DropdownContainer>
-                  {props.options.map((item, index) => {
+                  {props.options.map((item: TItem, index: number) => {
                     const itemProps = getItemProps({
                       index,
                       item,
                     });
                     const dropdownItemProps = itemProps;
-                    const isSelected = getIsSelected(props, item);
+                    const isSelected = getIsSelected(props, item as TItem);
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.tsx:117]: I'm curious about typing the `item` here as we already typed it in the `map` function definition.

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx [hunk 6]
```diff
@@ -36,10 +41,115 @@ import { MARK_TAGS, BLOCK_TAGS } from '../tags';
...
+type TRef = {
+  registerContentNode: TNodeRefObject;
+  containerRef?: LegacyRef<HTMLDivElement>;
+};
+
+type TType = {
+  type: string;
+};
+
+type TEditor = {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx:82]: I noticed we are using the editor in a bunch of this `rich-text-utils` files and we're typing it everywhere is needed (only with the slice of attributes used in each context).

I was wondering if it would be more maintainable if we have a shared types in this package where we precisely type the editor and use that type everywhere across this package files.

**Review by Rhotimee — CHANGES_REQUESTED** [review]
The files in the html folder hasn't been converted to typescript yet.
`packages/components/inputs/rich-text-utils/src/html`

File: packages/components/inputs/rich-text-utils/src/plugins/mark.tsx [hunk 7]
```diff
@@ -35,20 +69,26 @@ const MarkPlugin = (options = {}) => {
...
+      ): void {
+        if (!isHotKey(event as KeyboardEvent)) {
           return next();
         }
 
         event.preventDefault();
         editor.toggleMark(options.typeName);
       },
-      renderMark(props, editor, next) {
+      //@ts-ignore
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/plugins/mark.tsx:84]: why are not replace editor with an underscore and remove the `ts-ignore`

File: packages/components/inputs/rich-text-utils/src/plugins/placeholder.tsx [hunk 8]
```diff
@@ -1,6 +1,54 @@
 // copied from https://github.com/ianstormtaylor/slate/tree/master/packages/slate-react-placeholder/src
 // and modified to use editor's `options.placeholder` prop instead.
+//@ts-nocheck
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/plugins/placeholder.tsx:3]: Is there a reason we are not type checking the entire file?

File: packages/components/inputs/rich-text-utils/src/plugins/plugins.tsx [hunk 9]
```diff
@@ -1,9 +1,22 @@
...
+  attributes: unknown;
+  node: TType;
+};
+
 // eslint-disable-next-line import/prefer-default-export
 export const RenderBlockPlugin = () => {
   return {
-    renderBlock(props, editor, next) {
+    //@ts-ignore
+    renderBlock(props: TProps, editor: unknown, next: () => ReactNode) {
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/plugins/plugins.tsx:19]: replace editor with an underscore and remove `\\@ts-ignore`

File: packages/components/inputs/rich-text-utils/src/plugins/plugins.tsx [hunk 10]
```diff
@@ -41,7 +54,8 @@ export const RenderBlockPlugin = () => {
 
 export const RenderMarkPlugin = () => {
   return {
-    renderMark(props, editor, next) {
+    //@ts-ignore
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/plugins/plugins.tsx:57]: replace editor with an underscore and remove `\\@ts-ignore`

File: packages/components/inputs/rich-text-utils/src/plugins/mark.tsx (hunk 7 above)
Comment (Rhotimee) [reply packages/components/inputs/rich-text-utils/src/plugins/mark.tsx:84]: If it is possible and if it won't break anything, It'll be nice to remove the editor parameter and arguments in the renderMark function and wherever it is used.

File: packages/components/inputs/rich-text-utils/src/plugins/undo.ts [hunk 11]
```diff
@@ -1,25 +1,36 @@
+import { KeyboardEvent } from 'react';
+//@ts-ignore
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/plugins/undo.ts:2]: why do we have this here?

File: packages/components/inputs/rich-text-utils/src/plugins/undo.ts [hunk 12]
```diff
@@ -1,25 +1,36 @@
...
 import { isKeyHotkey } from 'is-hotkey';
 import memoize from 'lodash/memoize';
+import { type TEditor, type TOptions } from './redo';
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/plugins/undo.ts:5]:
```suggestion
import type { TEditor, TOptions } from './redo';
```

File: packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.styles.tsx [hunk 13]
```diff
@@ -2,6 +2,15 @@ import { css } from '@emotion/react';
 import styled from '@emotion/styled';
 import { customProperties as vars } from '@commercetools-uikit/design-system';
 
+type Props = {
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.styles.tsx:5]: I think it will be nice to rename this to `TDropdownItemProps` or something like that. Depending on how it is used.

File: packages/components/inputs/rich-text-utils/src/rich-text-body/hidden-input.tsx [hunk 14]
```diff
@@ -2,7 +2,15 @@ import { useCallback } from 'react';
 import PropTypes from 'prop-types';
 import { accessibleHiddenInputStyles } from '@commercetools-uikit/input-utils';
 
-const HiddenInput = (props) => {
+type Props = {
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/hidden-input.tsx:5]: rename Props to something more specific

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body-button.tsx [hunk 15]
```diff
@@ -1,16 +1,25 @@
-import PropTypes from 'prop-types';
 import omit from 'lodash/omit';
 import { css } from '@emotion/react';
 import { customProperties as vars } from '@commercetools-uikit/design-system';
+import { type ReactNode } from 'react';
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body-button.tsx:4]:
```suggestion
import type { ReactNode } from 'react';
```

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body-button.tsx [hunk 16]
```diff
@@ -1,16 +1,25 @@
...
+
+type Props = {
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body-button.tsx:6]: rename Props to something more specific

Comment (Rhotimee) [reply packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body-button.tsx:6]: Check other files also. The same was done in some other files.

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts [hunk 17]
```diff
@@ -28,17 +35,18 @@ const getBorderColor = (props) => {
...
     return css`
       background-color: ${vars.backgroundColorForInputWhenDisabled};
     `;
   }
-  if (props.isReadOnly) {
-    return css`
-      background-color: ${vars.backgroundColorForInputWhenReadonly};
-    `;
-  }
+  //Question: imported vars has no property with `backgroundColorForInputWhenReadonly` as key. Should this be defined?
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts:44]: Did you forget to uncomment this?

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx [hunk 18]
```diff
@@ -462,18 +572,6 @@ const RichTextEditorBody = forwardRef((props, ref) => {
...
-RichTextEditorBody.propTypes = {
-  hasError: PropTypes.bool,
-  hasWarning: PropTypes.bool,
-  isReadOnly: PropTypes.bool,
-  isDisabled: PropTypes.bool,
-  editor: PropTypes.any,
-  children: PropTypes.node,
-  containerStyles: PropTypes.any,
-  showExpandIcon: PropTypes.bool.isRequired,
-  onClickExpand: requiredIf(PropTypes.func, (props) => props.showExpandIcon),
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx:474]: It'll be nice to trigger a warning if props.showExpandIcon is true and onClickExpand is not a function

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx [hunk 19]
```diff
@@ -462,18 +572,6 @@ const RichTextEditorBody = forwardRef((props, ref) => {
...
-  styles: PropTypes.any,
-};
 
 RichTextEditorBody.defaultProps = {
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx:576]: Please type the defaultProps

File: packages/components/inputs/rich-text-utils/src/rich-text-plugins.tsx [hunk 20]
```diff
@@ -35,7 +58,7 @@ const plugins = [
           isParagraph;
 
         // eslint-disable-next-line no-param-reassign
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/rich-text-plugins.tsx:60]: Please, remove all unnecessary eslint comments

File: packages/components/inputs/rich-text-utils/src/rich-text-plugins.tsx [hunk 21]
```diff
@@ -5,24 +5,47 @@ import MarkPlugin from './plugins/mark';
 import { RenderMarkPlugin, RenderBlockPlugin } from './plugins';
 import PlaceholderPlugin from './plugins/placeholder';
 import { BLOCK_TAGS, MARK_TAGS } from './tags';
+import { ReactNode } from 'react';
```
Comment (Rhotimee) [inline packages/components/inputs/rich-text-utils/src/rich-text-plugins.tsx:8]:
```suggestion
import type { ReactNode } from 'react';
```

**Review by Rhotimee — CHANGES_REQUESTED** [review]

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx (hunk 6 above)
Comment (ddouglasz) [reply packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx:82]: That's a good idea. Will do that.

File: packages/components/inputs/rich-text-utils/src/plugins/placeholder.tsx [hunk 22]
```diff
@@ -1,6 +1,54 @@
...
+    };
+  };
+};
+
+type TData = {
+  get: (key: string) => number;
+};
+
+type TRenderDecorationProps = {
+  style: unknown;
```
[the remark this answers is not in the document]
Comment (ddouglasz) [reply packages/components/inputs/rich-text-utils/src/plugins/placeholder.tsx:45]: Thank you for the suggestion.

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts (hunk 17 above)
Comment (ddouglasz) [reply packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts:44]: Yes I forgot. lol, thank you. Will clean this up.

File: packages/components/inputs/rich-text-utils/src/plugins/placeholder.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply packages/components/inputs/rich-text-utils/src/plugins/placeholder.tsx:45]: Actually, I have a proposal for refactoring the types so we use more meaningful names and specific types like the one mentioned by @Rhotimee 

Let me know what you think about this:
```

type TEditorQueryPredicate = string | (() => void);

type TDecorationNode = {
  key: string;
  offset: number;
  path: string;
};

type TDecoration = {
  type: string;
  data: {
    key: number,
    get: (name: string) => number
  };
  anchor: TDecorationNode;
  focus: TDecorationNode;
};

type TNextDecorators = () => TDecoration[];

type TNodeTextsResult = [
  { key: string; text: string; },
  string
];

type TNode = {
  texts: (s?: Record<string, string>) => TNodeTextsResult[];
};

type TEditor = {
  query: (
    when: TEditorQueryPredicate,
    node: TNode
  ) => {
    next: () => void;
  };
  props: {
    options: {
      placeholder: string;
    };
  };
};

type TRenderDecorationProps = {
  children: ReactNode;
  decoration: TDecoration;
};

type TSlateReactPlaceholderOptions = {
  when: TEditorQueryPredicate;
  style: CSSProperties
};
```

File: packages/components/inputs/rich-text-utils/src/plugins/mark.tsx [hunk 23]
```diff
@@ -34,21 +67,25 @@ const MarkPlugin = (options = {}) => {
 
   return [
     {
-      // eslint-disable-next-line consistent-return
-      onKeyDown(event, editor, next) {
-        if (!isHotKey(event)) {
+      onKeyDown(
+        event: TEvent,
+        editor: { toggleMark: (arg0: unknown) => void },
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/plugins/mark.tsx:72]: Could we use the `TEditor` type for this `editor` parameter?

File: packages/components/inputs/rich-text-utils/src/plugins/mark.tsx [hunk 24]
```diff
@@ -34,21 +67,25 @@ const MarkPlugin = (options = {}) => {
...
+      ): void {
+        if (!isHotKey(event as KeyboardEvent)) {
           return next();
         }
 
         event.preventDefault();
         editor.toggleMark(options.typeName);
       },
-      renderMark(props, editor, next) {
+      renderMark(props: TProps, next: () => ReactJSXElement) {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/plugins/mark.tsx:82]: Here we're changing the function signature and I'm wondering if all consumers have also been adapted to it or checked that it won't affect them.

Comment (ddouglasz) [reply packages/components/inputs/rich-text-utils/src/plugins/mark.tsx:82]: hmm, I get you here. I might just see if I can make it uniform else I would restore to the initial function siignature and use `_` on the `editor` parameter since it's mostly not even used in the function.

File: packages/components/inputs/rich-text-utils/src/plugins/placeholder.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (emmenko) [reply packages/components/inputs/rich-text-utils/src/plugins/placeholder.tsx:45]: Agreed, we should try to use more clear names.

File: packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.tsx [hunk 25]
```diff
@@ -10,26 +14,69 @@ import {
   DropdownItem as StyledDropdownItem,
 } from './dropdown.styles';
 
-const getIsSelected = (props, item) =>
+type Props = {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.tsx:17]: I guess it would be better to name this following the name of the component: `TDropdownProps`.

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts [hunk 26]
```diff
@@ -2,7 +2,14 @@ import styled from '@emotion/styled';
 import { css } from '@emotion/react';
 import { customProperties as vars } from '@commercetools-uikit/design-system';
 
-const getBorderColor = (props) => {
+type Props = {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts:5]: Maybe we can pick props from the `TRichTextEditorBody` type defined in the `rich-text-body.tsx` file and name this one like `TRichTextBodyStylesProps`.
What do you think about it?

File: packages/components/inputs/rich-text-utils/src/plugins/placeholder.tsx [hunk 27]
```diff
@@ -1,6 +1,37 @@
 // copied from https://github.com/ianstormtaylor/slate/tree/master/packages/slate-react-placeholder/src
 // and modified to use editor's `options.placeholder` prop instead.
 import { warning } from '@commercetools-uikit/utils';
+import { ReactJSXElement } from '@emotion/react/types/jsx-namespace';
+import { type CSSProperties, ReactNode } from 'react';
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/plugins/placeholder.tsx:5]:
```suggestion
import type { CSSProperties, ReactNode } from 'react';
```

File: packages/components/inputs/rich-text-utils/src/plugins/redo.ts [hunk 28]
```diff
@@ -1,42 +1,47 @@
+import { KeyboardEvent } from 'react';
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/plugins/redo.ts:1]:
```suggestion
import type { KeyboardEvent } from 'react';
```

File: packages/components/inputs/rich-text-utils/src/plugins/plugins.tsx [hunk 29]
```diff
@@ -1,9 +1,20 @@
+import { ReactNode } from 'react';
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/plugins/plugins.tsx:1]:
```suggestion
import type { ReactNode } from 'react';
```

File: packages/components/inputs/rich-text-utils/src/plugins/redo.ts [hunk 30]
```diff
@@ -1,42 +1,47 @@
...
+        editor: TEditor,
+        next: () => void
+      ) {
         if (!isRedoHotkey(event) || !hasRedos(editor)) {
           return next();
         }
 
         event.preventDefault();
-        editor.toggleRedo();
+        editor.toggleRedo!();
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/plugins/redo.ts:33]: Is this correct? Or did you mean this?

```suggestion
        editor.toggleRedo?.();
```

File: packages/components/inputs/rich-text-utils/src/plugins/redo.ts [hunk 31]
```diff
@@ -1,42 +1,47 @@
...
       },
       commands: {
-        toggleRedo: (editor) => {
+        toggleRedo: (editor: TEditor) => {
           if (!editor.value.selection.isFocused) {
             editor.focus();
           }
           if (hasRedos(editor)) {
-            editor.redo();
+            editor.redo!();
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/plugins/redo.ts:44]: Same here.

File: packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.tsx [hunk 32]
```diff
@@ -1,7 +1,11 @@
-import Downshift from 'downshift';
+/// <reference types="@emotion/react/types/css-prop" />
+import { ElementType, FunctionComponent } from 'react';
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.tsx:2]:
```suggestion
import type { ElementType, FunctionComponent } from 'react';
```

File: packages/components/inputs/rich-text-utils/src/rich-text-body/hidden-input.tsx [hunk 33]
```diff
@@ -1,8 +1,15 @@
 import { useCallback } from 'react';
-import PropTypes from 'prop-types';
 import { accessibleHiddenInputStyles } from '@commercetools-uikit/input-utils';
 
-const HiddenInput = (props) => {
+type Props = {
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/hidden-input.tsx]:
```suggestion
type THiddenInputProps = {
```

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts [hunk 34]
```diff
@@ -1,8 +1,14 @@
 import styled from '@emotion/styled';
 import { css } from '@emotion/react';
 import { customProperties as vars } from '@commercetools-uikit/design-system';
+import type { TRichTextEditorBody } from './rich-text-body';
 
-const getBorderColor = (props) => {
+type TRichTextBodyStylesProps = Pick<
+  TRichTextEditorBody,
+  'hasError' | 'isReadOnly' | 'hasWarning' | 'isReadOnly' | 'isDisabled'
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts:8]:
```suggestion
  'hasError' | 'isReadOnly' | 'hasWarning' | 'isDisabled'
```

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts [hunk 35]
```diff
@@ -28,17 +34,12 @@ const getBorderColor = (props) => {
...
   if (props.isDisabled) {
     return css`
       background-color: ${vars.backgroundColorForInputWhenDisabled};
     `;
   }
-  if (props.isReadOnly) {
-    return css`
-      background-color: ${vars.backgroundColorForInputWhenReadonly};
-    `;
-  }
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts:41]: Why is this removed?

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx [hunk 36]
```diff
@@ -168,8 +220,8 @@ const createMoreStylesDropdownOptions = (intl) => {
   ];
 };
 
-const RichTextEditorBody = forwardRef((props, ref) => {
-  const { registerContentNode, containerRef } = ref;
+const RichTextEditorBody = forwardRef((props: TRichTextEditorBody, ref) => {
+  const { registerContentNode, containerRef } = ref as unknown as TRef;
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx:224]:
```suggestion
const RichTextEditorBody = forwardRef<TRef, TRichTextEditorBody>((props, ref) => {
  const { registerContentNode, containerRef } = ref;
```

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx [hunk 37]
```diff
@@ -222,24 +276,26 @@ const RichTextEditorBody = forwardRef((props, ref) => {
 
   const onChangeMoreStyles = useCallback(
     (val) => {
-      props.editor.toggleMark(val.value);
+      props.editor.toggleMark!(val.value);
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx:279]: Maybe this?

```suggestion
      props.editor.toggleMark?.(val.value);
```

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx [hunk 38]
```diff
@@ -222,24 +276,26 @@ const RichTextEditorBody = forwardRef((props, ref) => {
...
-    (props.editor.value.blocks.first() &&
-      props.editor.value.blocks.first().type) ||
+    (props.editor.value?.blocks.first!() &&
+      props.editor.value?.blocks.first!().type) ||
     '';
 
   // so that we don't show our multi dropdown in an `indeterminate`
   // while the component is not in focus
-  let activeMoreStyleMarks = [];
+  let activeMoreStyleMarks: Array<unknown> = [];
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx:291]:
```suggestion
  let activeMoreStyleMarks: Array<string> = [];
```

File: packages/components/inputs/rich-text-utils/src/editor.types.ts [hunk 39]
```diff
@@ -0,0 +1,114 @@
...
+
+export type TNode = {
+  texts: (nodeParam?: Record<string, string>) => TNodeTextsResult[];
+};
+
+export type TOptions = {
+  hotkey: string;
+};
+
+export type TEditor = {
```
Comment (emmenko) [inline packages/components/inputs/rich-text-utils/src/editor.types.ts:72]: Do we need to define the types for this manually? Isn't there a type exposed from the library? 🤔

**Review by emmenko — CHANGES_REQUESTED** [review]
Thanks for fighting through this!

Definitely not the easiest package to migrate 😅

I left some comments around types but in general it looks good.

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts (hunk 35 above)
Comment (ddouglasz) [reply packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.styles.ts:41]: `backgroundColorForInputWhenReadonly` does not exist on from  `@commercetools-uikit/design-system` where `vars` is coming from.

File: packages/components/inputs/rich-text-utils/src/html/html.tsx [hunk 40]
```diff
@@ -1,7 +1,15 @@
-/* eslint-disable consistent-return */
-import Html from 'slate-html-serializer';
+import Html, { type HtmlOptions } from 'slate-html-serializer';
 import flatMap from 'lodash/flatMap';
 import { MARK_TAGS, BLOCK_TAGS } from '../tags';
+import type { ReactNode } from 'react';
+
+type TSerializeObject = {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/html/html.tsx:6]: Instead of a verb, maybe it's better to use a noun for types.
Perhaps we could use here something like `TSerializableObject`.

File: packages/components/inputs/rich-text-utils/src/localized/localized.ts [hunk 41]
```diff
@@ -0,0 +1,64 @@
+import { warning } from '@commercetools-uikit/utils';
+import uniq from 'lodash/uniq';
+import html from '../html';
+import isRichTextEmpty from '../is-empty';
+
+type TLocalizedString = {
+  language: string;
+};
+
+type TTranslations = {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/localized/localized.ts:10]: I've seen that we're using `Record<string, string>` is other parts of the repository for this case:
```
type TTranslations = Record<string, string>;
```

File: packages/components/inputs/rich-text-utils/src/localized/localized.ts [hunk 42]
```diff
@@ -0,0 +1,64 @@
+import { warning } from '@commercetools-uikit/utils';
+import uniq from 'lodash/uniq';
+import html from '../html';
+import isRichTextEmpty from '../is-empty';
+
+type TLocalizedString = {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/localized/localized.ts:6]: I guess this type would shape objects like this:
```
{
  messages: 'en'
}
```

But I think that, in the context it is used, the expected object is like this:
```
{
  'en': 'Hello',
  'es': 'Hola'
}
```
so the type should be:
```
type LocalizedString = {
  [language: string]: string;
}

// OR
type TLocalizedString = Record<string, string>;
```

What do you think?

File: packages/components/inputs/rich-text-utils/src/plugins/mark.tsx [hunk 43]
```diff
@@ -1,6 +1,25 @@
...
+type TMarkPluginOptions = {
+  hotkey: string;
+  typeName: string;
+  RenderMark: ReactNode;
+  command: string;
+  query: string;
+  [option: string]: string | ReactNode;
+};
+
+type TProps = {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/plugins/mark.tsx:16]:
```suggestion
type TRenderMarkProps = {
```
Maybe we could use a more specific name.

File: packages/components/inputs/rich-text-utils/src/editor.types.ts [hunk 44]
```diff
@@ -0,0 +1,114 @@
...
+
+export type TMark = {
+  type: string;
+};
+
+export type TNode = {
+  texts: (nodeParam?: Record<string, string>) => TNodeTextsResult[];
+};
+
+export type TOptions = {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/editor.types.ts:68]: I feel this name is too abstract. What are this options for?
Maybe we can find something more related to what it is used for.

**Review by CarlosCortizasCT — CHANGES_REQUESTED** [review]

File: packages/components/inputs/rich-text-utils/src/localized/localized.ts (hunk 42 above)
Comment (ddouglasz) [reply packages/components/inputs/rich-text-utils/src/localized/localized.ts:6]: This makes sense. Thanks for pointing it out.

File: packages/components/inputs/rich-text-utils/src/editor.types.ts [hunk 45]
```diff
@@ -0,0 +1,114 @@
...
+
+export type TMark = {
+  type: string;
+};
+
+export type TNode = {
+  texts: (nodeParam?: Record<string, string>) => TNodeTextsResult[];
+};
+
+export type THotKeyptions = {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/editor.types.ts:68]: There is a typo here: `THotKeyptions` -> `THotKeyOptions`

File: packages/components/inputs/rich-text-utils/src/localized/localized.ts [hunk 46]
```diff
@@ -0,0 +1,63 @@
+import { warning } from '@commercetools-uikit/utils';
+import uniq from 'lodash/uniq';
+import html from '../html';
+import isRichTextEmpty from '../is-empty';
+
+type TLocalizedString = Record<string, string>;
+// type TLocalizedString = {
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/localized/localized.ts:7]: We should remove this commented code.

**Review by CarlosCortizasCT — CHANGES_REQUESTED** [review]

File: packages/components/inputs/rich-text-utils/src/editor.types.ts (hunk 39 above)
Comment (kark) [reply packages/components/inputs/rich-text-utils/src/editor.types.ts:72]: If I may: it seems that `Editor` class from https://www.npmjs.com/package/@types/slate-react/v/0.22.9 is the closest to what we expect from `TEditor` type, but if I'm not mistaken this class would have to be extended manually anyway 🤔

Comment (emmenko) [reply packages/components/inputs/rich-text-utils/src/editor.types.ts:72]: So Slate has [TS support](https://docs.slatejs.org/concepts/12-typescript) but not in the current version we are using.

I opened a separate issue for that: #2120 

So let's keep the "custom types" for now.

**Comment by emmenko** [comment]
I pushed some commits for cleaning up some leftovers. I found it easier than keep adding comments.

I think this should be more or less good to go now.

**Comment by ddouglasz** [comment]
> I pushed some commits for cleaning up some leftovers. I found it easier than keep adding comments.
> 
> I think this should be more or less good to go now.

Thank you.

File: packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx [hunk 47]
```diff
@@ -87,23 +143,18 @@ const Item = styled.div`
...
 `;
 
-const StylesDropdownItem = (props) => {
-  const as =
-    Object.keys(BLOCK_TAGS).find((key) => BLOCK_TAGS[key] === props.value) ||
-    'div';
+const StylesDropdownItem = (props: TStylesDropdownItem) => {
+  const asProp = (Object.keys(BLOCK_TAGS).find(
+    (key) => BLOCK_TAGS[key as keyof typeof BLOCK_TAGS] === props.value
+  ) || 'div') as React.ElementType;
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx:149]: nit: Add `ElementType` to the rest of React imports at the beginning of the file.

**Review by CarlosCortizasCT — APPROVED** [review]
LGTM

**Review by emmenko — APPROVED** [review]
Alright, this is good enough now. Thanks for keeping up with all the feedback and changes, this was definitely one of the more complex packages to migrate 😅

**Review by Rhotimee — APPROVED** [review]
Great job @ddouglasz

**Comment by CarlosCortizasCT** [comment]
@ddouglasz you rock, man! 🚀
